analytics/tracker.go

```diff
@@ -0,0 +1,226 @@
+package analytics
+
+import (
+	"bytes"
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/dagger/dagger/core/pipeline"
+	"github.com/dagger/dagger/engine"
+)
+
+const (
+	flushInterval = 1 * time.Second
+	trackURL      = "https://analytics.dagger.cloud/track"
+)
+
+type Event struct {
+	Timestamp  time.Time              `json:"ts,omitempty"`
+	Type       string                 `json:"type,omitempty"`
+	Properties map[string]interface{} `json:"properties,omitempty"`
+
+	DeviceID string `json:"device_id,omitempty"`
+	OrgID    string `json:"org_id,omitempty"`
+	IP       string `json:"ip,omitempty"`
+
+	ClientVersion string `json:"client_version,omitempty"`
+	ClientOS      string `json:"client_os,omitempty"`
+	ClientArch    string `json:"client_arch,omitempty"`
+
+	CI       bool   `json:"ci"`
+	CIVendor string `json:"ci_vendor,omitempty"`
+
+	GitRemoteEncoded string `json:"git_remote_encoded,omitempty"`
+}
+
+type Tracker interface {
+	Capture(event string, properties map[string]any)
+	io.Closer
+}
+
+type analyticsKey struct{}
+
+func WithContext(ctx context.Context, t Tracker) context.Context {
+	return context.WithValue(ctx, analyticsKey{}, t)
+}
+
+func Ctx(ctx context.Context) Tracker {
+	if t := ctx.Value(analyticsKey{}); t != nil {
+		return t.(Tracker)
+	}
+	return &noopTracker{}
+}
+
+type noopTracker struct {
+}
+
+func (t *noopTracker) Capture(event string, properties map[string]any) {
+	// FIXME: remove this
+	fmt.Fprintf(os.Stderr, "NO OP PUSH!!\n")
+}
+
+func (t *noopTracker) Close() error {
+	return nil
+}
+
+type CloudTracker struct {
+	labels     map[string]string
+	cloudToken string
+
+	closed bool
+	mu     sync.Mutex
+	queue  []*Event
+	stopCh chan struct{}
+	doneCh chan struct{}
+}
+
+func New(labels ...pipeline.Label) Tracker {
+	// from https://consoledonottrack.com/
+	if v := os.Getenv("DO_NOT_TRACK"); v == "1" {
```

Comment:
Since `CloudTracker` is used both client and server side, in order for it to be effective a user would need to set it both places IIUC?

I wonder if we should have it exclusively set client-side and then tell the server the setting so it only needs to be set one place by the client? If so, could just add a field to `engine.ClientMetadata`, set the field [here](https://github.com/sipsma/dagger/blob/c5eaf680bb1d888ab17f866fb5881196d7767a65/engine/client/client.go#L286-L286) and give it to the server from `opts` [here](https://github.com/sipsma/dagger/blob/c5eaf680bb1d888ab17f866fb5881196d7767a65/engine/server/buildkitcontroller.go#L270-L270).

Reply:
IMO, it would be useful if we could support both - if it's set in the client *or* the server, telemetry is disabled.

e.g. as a end-user, I don't want my usage to be tracked. But also, as an admin, I might want to deploy dagger engine in a way that does not leak info outside my org.

Reply:
👍 Done; supports both